code/game/machinery/doors/firedoor.dm

```diff
@@ -203,30 +198,59 @@
 		for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
 			fire_panel.set_status()
 
-/obj/machinery/door/firedoor/proc/check_atmos(datum/source)
+/obj/machinery/door/firedoor/proc/refresh_shared_turfs(datum/source, list/leaving_members, list/joining_members)
+	SIGNAL_HANDLER
+	
+	for(var/obj/machinery/door/firedoor/firelock in my_area.firedoors)
+		firelock.check_adjacent_turfs()
+
+/obj/machinery/door/firedoor/proc/check_adjacent_turfs()
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(src),dir)
+		if(checked_turf && !checked_turf.density)
+			RegisterSignal(checked_turf, COMSIG_TURF_EXPOSE, .proc/process_results, override = TRUE)
+			RegisterSignal(checked_turf, COMSIG_TURF_CALCULATED_ADJACENT_ATMOS, .proc/process_results, override = TRUE)
+
+/obj/machinery/door/firedoor/proc/check_atmos(turf/checked_turf)
+	var/datum/gas_mixture/environment = checked_turf.return_air()
+	if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST && my_area.active_alarms[ALARM_ATMOS])
+		return FIRELOCK_ALARM_TYPE_HOT
+	if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT && my_area.active_alarms[ALARM_ATMOS])
+		return FIRELOCK_ALARM_TYPE_COLD
+	return
+
+/obj/machinery/door/firedoor/proc/process_results(datum/source)
+	SIGNAL_HANDLER
+
 	if(!COOLDOWN_FINISHED(src, detect_cooldown))
 		return
-	if(alarm_type)
+	if(alarm_type == FIRELOCK_ALARM_TYPE_GENERIC)
 		return
+
 	for(var/area/place in affecting_areas)
 		if(!place.fire_detect) //if any area is set to disable detection
 			return
 
-	var/turf/my_turf = source
-	if(!my_turf)
-		my_turf = get_turf(src)
-
-	var/datum/gas_mixture/environment = my_turf.return_air()
-	var/result
-
-	if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
-		result = FIRELOCK_ALARM_TYPE_HOT
-	if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
-		result = FIRELOCK_ALARM_TYPE_COLD
-	if(!result)
+	var/turf/checked_turf = source
+	var/datum/merger/merge_group = GetMergeGroup(merger_id, merger_typecache)
+	var/result = check_atmos(checked_turf)
+	
+	if(!issue_turfs?.len && alarm_type)
+		start_deactivation_process()
 		return
-
-	start_activation_process(result)
+	else if(issue_turfs?.len && !alarm_type)
+		result = check_atmos(issue_turfs[1])
+		if(result)
+			start_activation_process(result)
+			return
+	if(result && TURF_SHARES(checked_turf))
+		for(var/obj/machinery/door/firedoor/firelock as anything in merge_group.members)
+			if(!firelock.issue_turfs.Find(checked_turf))
+				firelock.issue_turfs |= checked_turf
```

Comment:
Sorry I should have been more clear.
|= essentially says if the thing is not already in this list, insert it into the list.
That's why I suggested it, so you wouldn't need .Find()